In [47]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import xgboost as xgb
import seaborn as sb
from sklearn.metrics import auc, accuracy_score, confusion_matrix, mean_squared_error
from sklearn.model_selection import cross_val_score, GridSearchCV, KFold, RandomizedSearchCV, train_test_split
from sklearn.metrics import classification_report
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import f1_score, precision_score, recall_score
from sklearn import preprocessing
from sklearn import linear_model
from sklearn.metrics import r2_score, mean_squared_error
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [48]:
os.chdir('/content/drive/My Drive/Bootcamp ML2020/')
os.listdir('.')
dataset = pd.read_csv('heart.csv')
dataset.head()

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]

In [49]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [50]:
dataset.describe()

age         sex          cp  ...          ca        thal      target
count  303.000000  303.000000  303.000000  ...  303.000000  303.000000  303.000000
mean    54.366337    0.683168    0.966997  ...    0.729373    2.313531    0.544554
std      9.082101    0.466011    1.032052  ...    1.022606    0.612277    0.498835
min     29.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000
25%     47.500000    0.000000    0.000000  ...    0.000000    2.000000    0.000000
50%     55.000000    1.000000    1.000000  ...    0.000000    2.000000    1.000000
75%     61.000000    1.000000    2.000000  ...    1.000000    3.000000    1.000000
max     77.000000    1.000000    3.000000  ...    4.000000    3.000000    1.000000

[8 rows x 14 columns]

In [51]:
dataset['target'].value_counts()

1    165
0    138
Name: target, dtype: int64

In [52]:
dataset.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

In [53]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [54]:
# sb.set_style("whitegrid");
# sb.pairplot(dataset, hue="target", size=3);
# plt.show()

In [55]:
Y = dataset['target'] # Target prediksi
x = dataset.drop(['target'], axis=1) # Data Id tidak diperlukan

In [56]:
scaler = StandardScaler()
X = scaler.fit_transform(x)
data_x = pd.DataFrame(X, columns=x.columns)

In [57]:
x_train, x_test, y_train, y_test = train_test_split(data_x, Y, test_size=0.2)

In [58]:
from imblearn.over_sampling import SMOTE
smt=SMOTE()
x_train, y_train = smt.fit_sample(x_train, y_train)
np.bincount(y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function safe_indexing is deprecated; safe_indexing is deprecated in version 0.22 and will be removed in version 0.24.
  warnings.warn(msg, category=FutureWarning)


array([131, 131])

**Logistic Regression**

In [59]:
logreg = LogisticRegression()
logreg.fit(x_train, y_train)
y_pred = logreg.predict(x_test)
print('Accuracy of logistic regression classifier on train set: {:.2f}'.format(logreg.score(x_train, y_train))) # .score() -> Accuracy
print('Accuracy of logistic regression classifier on test set: {:.2f}'.format(logreg.score(x_test, y_test))) # .score() -> Accuracy
print("f1 score         :",f1_score(y_test, y_pred, average='macro'))
print("precision score  :",precision_score(y_test, y_pred, average='macro'))
print("recall score     :",recall_score(y_test, y_pred, average='macro'))

Accuracy of logistic regression classifier on train set: 0.85
Accuracy of logistic regression classifier on test set: 0.84
f1 score         : 0.83495670995671
precision score  : 0.834051724137931
recall score     : 0.8376906318082789


In [60]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.79      0.85      0.82        27
           1       0.88      0.82      0.85        34

    accuracy                           0.84        61
   macro avg       0.83      0.84      0.83        61
weighted avg       0.84      0.84      0.84        61



**Decision Tree Classifier**

In [61]:
from sklearn.tree import DecisionTreeClassifier
dectree = DecisionTreeClassifier()
dectree.fit(x_train, y_train)
y_pred = dectree.predict(x_test)
print('Accuracy of Decision Tree Classifier on train set: {:.2f}'.format(dectree.score(x_train, y_train))) # .score() -> Accuracy
print('Accuracy of Decision Tree Classifier on test set: {:.2f}'.format(dectree.score(x_test, y_test))) # .score() -> Accuracy
print("f1 score         :",f1_score(y_test, y_pred, average='macro'))
print("precision score  :",precision_score(y_test, y_pred, average='macro'))
print("recall score     :",recall_score(y_test, y_pred, average='macro'))


Accuracy of Decision Tree Classifier on train set: 1.00
Accuracy of Decision Tree Classifier on test set: 0.77
f1 score         : 0.768939393939394
precision score  : 0.7683189655172413
recall score     : 0.7712418300653594


In [62]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.72      0.78      0.75        27
           1       0.81      0.76      0.79        34

    accuracy                           0.77        61
   macro avg       0.77      0.77      0.77        61
weighted avg       0.77      0.77      0.77        61



**SVM**


In [71]:
from sklearn.svm import SVC

def kernel_svc(ks):
  classifier_svc = SVC(kernel = ks, shrinking=True)
  classifier_svc.fit(x_train, y_train)
  y_pred = classifier_svc.predict(x_test)
  print('Accuracy of SVC ',ks,' classifier on test set: {:.2f}'.format(classifier_svc.score(x_train, y_train))) # .score() -> Accuracy
  print('Accuracy of SVC ',ks,' classifier on test set: {:.2f}'.format(classifier_svc.score(x_test, y_test)))
  print(f1_score(y_test, y_pred, average="macro"))
  print(precision_score(y_test, y_pred, average="macro"))
  print(recall_score(y_test, y_pred, average="macro"))

kernelsvc=['rbf','poly','linear']
for x in kernelsvc:
  print("SVC linear: ", x)
  kernel_svc(x)
  print()

SVC linear:  rbf
Accuracy of SVC  rbf  classifier on test set: 0.92
Accuracy of SVC  rbf  classifier on test set: 0.87
0.8679653679653679
0.8669181034482758
0.8709150326797386

SVC linear:  poly
Accuracy of SVC  poly  classifier on test set: 0.93
Accuracy of SVC  poly  classifier on test set: 0.90
0.900974025974026
0.8997844827586207
0.9041394335511983

SVC linear:  linear
Accuracy of SVC  linear  classifier on test set: 0.84
Accuracy of SVC  linear  classifier on test set: 0.85
0.8510176390773405
0.850108225108225
0.85239651416122



In [64]:
classifier_linear_svc = SVC()
classifier_linear_svc.fit(x_train, y_train)
y_pred = classifier_linear_svc.predict(x_test)
print('Accuracy of LinearSVC classifier on train set: {:.2f}'.format(classifier_linear_svc.score(x_train, y_train))) # .score() -> Accuracy
print('Accuracy of LinearSVC classifier on test set: {:.2f}'.format(classifier_linear_svc.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

Accuracy of LinearSVC classifier on train set: 0.92
Accuracy of LinearSVC classifier on test set: 0.87
0.8679653679653679
0.8669181034482758
0.8709150326797386


**Random Forest**

In [65]:
from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(n_estimators=300, bootstrap=True, max_features = "sqrt")
rf_model.fit(x_train, y_train)
y_pred = rf_model.predict(x_test)
print('Accuracy of Random Forest on train set: {:.2f}'.format(rf_model.score(x_train, y_train)))
print('Accuracy of Random Forest on test set: {:.2f}'.format(rf_model.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

Accuracy of Random Forest on train set: 1.00
Accuracy of Random Forest on test set: 0.84
0.8338779956427015
0.8338779956427015
0.8338779956427015


**K-Nearest Neighbors**

In [66]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.neighbors import KNeighborsRegressor

def k_near(knnr):
  # knear = KNeighborsClassifier(knnr)
  # knear.fit(x_train, y_train)
  # y_pred = knear.predict(x_test)
  # print('Accuracy of KNN on train set: {:.2f}'.format(knear.score(x_train, y_train)))
  # print('Accuracy of KNN on test set: {:.2f}'.format(knear.score(x_test, y_test)))
  # print(f1_score(y_test, y_pred, average="macro"))
  # print(precision_score(y_test, y_pred, average="macro"))
  # print(recall_score(y_test, y_pred, average="macro"))
  
  knn = KNeighborsClassifier(knnr)
  knn.fit(x_train, y_train)
  f1_score(knn.predict(x_train), y_train)
  cross_validation_score = cross_val_score(knn, x_train, y_train, cv=10)
  knn_score = f1_score(knn.predict(x_test), y_test) 
  return knn_score

# k=[1,3,5,7,9,11,13,15,99]
# for x in k:
#   print("K Nearest Neighbor: ", x)
#   k_near(x)
#   print()

k=[1,3,5,7,9,11,13,15,99]
for x in k:
  print("K Nearest Neighbor: ", x)
  print(k_near(x))
  print()

K Nearest Neighbor:  1
0.8358208955223881

K Nearest Neighbor:  3
0.819672131147541

K Nearest Neighbor:  5
0.9090909090909091

K Nearest Neighbor:  7
0.90625

K Nearest Neighbor:  9
0.923076923076923

K Nearest Neighbor:  11
0.923076923076923

K Nearest Neighbor:  13
0.9393939393939394

K Nearest Neighbor:  15
0.8923076923076922

K Nearest Neighbor:  99
0.8857142857142858



**XGBoost**

In [67]:
xgb_model = xgb.XGBClassifier(objective="reg:linear")
xgb_model.fit(x_train, y_train)
y_pred = xgb_model.predict(x_test.values)
print('Accuracy of XGBoost on train set: {:.2f}'.format(xgb_model.score(x_train, y_train)))
print('Accuracy of XGBoost on test set: {:.2f}'.format(xgb_model.score(np.array(x_test), y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

[07:52:43] WARNING: /workspace/src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.
Accuracy of XGBoost on train set: 0.98
Accuracy of XGBoost on test set: 0.84
0.83495670995671
0.834051724137931
0.8376906318082789


**Gaussian Process Classifier**

In [68]:
from sklearn.gaussian_process import GaussianProcessClassifier


gaus_model = GaussianProcessClassifier()
gaus_model.fit(x_train, y_train)

y_pred = gaus_model.predict(x_test)
print('Accuracy of Gauss on train set: {:.2f}'.format(gaus_model.score(x_train, y_train)))
print('Accuracy of Gauss on test set: {:.2f}'.format(gaus_model.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

Accuracy of Gauss on train set: 1.00
Accuracy of Gauss on test set: 0.85
0.8523002421307506
0.8561422413793103
0.8600217864923747


**MLP Classifier**

In [69]:
from sklearn.neural_network import MLPClassifier
MLP_model = MLPClassifier(hidden_layer_sizes=(3,), activation='logistic',
                       solver='adam', alpha=0.0001,learning_rate='constant', 
                      learning_rate_init=0.3)
MLP_model.fit(x_train, y_train)

y_pred = MLP_model.predict(x_test)
print('Accuracy of MLP on train set: {:.2f}'.format(MLP_model.score(x_train, y_train)))
print('Accuracy of MLP on test set: {:.2f}'.format(MLP_model.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

Accuracy of MLP on train set: 0.89
Accuracy of MLP on test set: 0.89
0.884124830393487
0.8831168831168831
0.8856209150326797


**ADA Boost**

In [70]:
from sklearn.ensemble import AdaBoostClassifier
ada_model = AdaBoostClassifier(n_estimators=100, random_state=0)
ada_model.fit(x_train, y_train)

y_pred = ada_model.predict(x_test)
print('Accuracy of ADA Boost on train set: {:.2f}'.format(ada_model.score(x_train, y_train)))
print('Accuracy of ADA Boost on test set: {:.2f}'.format(ada_model.score(x_test, y_test)))
print(f1_score(y_test, y_pred, average="macro"))
print(precision_score(y_test, y_pred, average="macro"))
print(recall_score(y_test, y_pred, average="macro"))

Accuracy of ADA Boost on train set: 0.97
Accuracy of ADA Boost on test set: 0.80
0.802801724137931
0.8043010752688172
0.8082788671023965


Dari setiap algoritma yang saya gunakan, Algoritma yang memiliki tingkat akurasi tertinggi adalah algoritma k nearest neighbor dengan parameter yang di set k=13, yang memiliki tingkat akurasi 94%